### DISCUSSION TITLE
How do I write a Visitor method to insert a Field in a SelectionSet?

### DISCUSSION THREAD

**Discussion by richardkyee** [graphql-java/graphql-java#3633] [General]
I am trying to modify a GraphQL Document Query to insert a new field in the document by replacing a fieldA with fieldB which then contains fieldA.

I have tried using AstTransformer.transformer(document, visitor) to do the transform by overriding the
public TraversalControl visitField(Field field, TraverserContext<Node> data) method
and creating a new Field and then calling data.changeNode(newField) but the Document remains unchanged.
```
    private class CustomNodeVisitor extends NodeVisitorStub implements NodeVisitor {

        @Override
        public TraversalControl visitField(Field field, TraverserContext<Node> data) {
            if (field.getName().equals("fieldA")) {
                Field fieldB = Field.newField("fieldB").build();
                data.changeNode(connectionsField);
            } else if (field.getName().equals("connections")) {
                System.out.println("found connections Field");
return TraversalControl.CONTINUE;
        }
    }
 }
```

**Comment by richardkyee** [comment]
This above code is pseudo code. I have debugged the visitField method and it is getting called and the new field is getting traversed but the Node that is returned from the transform method is unchanged from the input.

**Reply by bbakerman** [reply]
```groovy
    def "one node changed"() {
        def document = TestUtil.parseQuery("{foo}")

        AstTransformer astTransformer = new AstTransformer()

        def visitor = new NodeVisitorStub() {

            @Override
            TraversalControl visitField(Field node, TraverserContext<Node> context) {
                Field changedField = node.transform({ builder -> builder.name("foo2") })
                return TreeTransformerUtil.changeNode(context, changedField)
            }
        }


        when:
        def newDocument = astTransformer.transform(document, visitor)

        then:
        printAstCompact(newDocument) == "{foo2}"

    }
```

from graphql.language.AstTransformerTest

The trick is to use a AST transformer and also to return via the  `graphql.util.TreeTransformerUtil#changeNode` method.

Is not intuitive since there is a `changeNode` on the context but it does not deep change things.   Rather you need to use `TreeTransformerUtil.changeNode`.

**Reply by richardkyee** [reply]
Thank you @bbakerman. I just also decided to check out the tests and also figured it out. I think I will do a PR and add some documentation to help others avode the hours of time spent trying to get it to work while using the wrong method.

**Reply by bbakerman** [reply]
Yeah it is tricky.  I am not sure what the `changeNode` on traversal context does honestly.  I never wrote any of this code.

@andimarek knows more

We do have https://www.graphql-java.com/documentation/schema#changing-schema - which is about changing schema but its the same generic visitor code mechanisms.

**Reply by dondonz** [reply]
If you'd like to help out with any documentation, our documentation repo is https://github.com/graphql-java/graphql-java-page and you're welcome to open a pull request

**Reply by richardkyee** [reply]
I will take a look and see what I can contribute. Thanks.
